#include "tabelaobjetoswidget.h"
#include "caixamensagem.h"
extern CaixaMensagem *caixa_msg;

TabelaObjetosWidget::TabelaObjetosWidget(unsigned conf_botoes, bool conf_exclusoes, QWidget *parent): QWidget(parent)
{
 setupUi(this);
 connect(mover_baixo_tb, SIGNAL(clicked(bool)), this, SLOT(moverLinhas(void)));
 connect(mover_cima_tb, SIGNAL(clicked(bool)), this, SLOT(moverLinhas(void)));
 connect(mover_primeiro_tb, SIGNAL(clicked(bool)), this, SLOT(moverLinhas(void)));
 connect(mover_ultimo_tb, SIGNAL(clicked(bool)), this, SLOT(moverLinhas(void)));
 connect(adicionar_tb, SIGNAL(clicked(bool)), this, SLOT(adicionarLinha(void)));
 connect(remover_tb, SIGNAL(clicked(bool)), this, SLOT(removerLinha(void)));
 connect(editar_tb, SIGNAL(clicked(bool)), this, SLOT(editarLinha(void)));
 connect(atualizar_tb, SIGNAL(clicked(bool)), this, SLOT(atualizarLinha(void)));
 connect(remover_todas_tb, SIGNAL(clicked(bool)), this, SLOT(removerLinhas(void)));
 connect(tabela_tbw, SIGNAL(cellClicked(int,int)), this, SLOT(habilitarBotoes(void)));
 connect(tabela_tbw, SIGNAL(cellDoubleClicked(int,int)), this, SLOT(editarLinha(void)));

 this->conf_exclusoes=conf_exclusoes;
 definirConfiguracaoBotoes(conf_botoes);

 definirNumColunas(1);

 adicionar_tb->setToolTip(adicionar_tb->toolTip() + QString(" (%1)").arg(adicionar_tb->shortcut().toString()));
 remover_tb->setToolTip(remover_tb->toolTip() + QString(" (%1)").arg(remover_tb->shortcut().toString()));
 remover_todas_tb->setToolTip(remover_todas_tb->toolTip() + QString(" (%1)").arg(remover_todas_tb->shortcut().toString()));
 atualizar_tb->setToolTip(atualizar_tb->toolTip() + QString(" (%1)").arg(atualizar_tb->shortcut().toString()));
 editar_tb->setToolTip(editar_tb->toolTip() + QString(" (%1)").arg(editar_tb->shortcut().toString()));
 mover_ultimo_tb->setToolTip(mover_ultimo_tb->toolTip() + QString(" (%1)").arg(mover_ultimo_tb->shortcut().toString()));
 mover_primeiro_tb->setToolTip(mover_primeiro_tb->toolTip() + QString(" (%1)").arg(mover_primeiro_tb->shortcut().toString()));
 mover_cima_tb->setToolTip(mover_cima_tb->toolTip() + QString(" (%1)").arg(mover_cima_tb->shortcut().toString()));
 mover_baixo_tb->setToolTip(mover_baixo_tb->toolTip() + QString(" (%1)").arg(mover_baixo_tb->shortcut().toString()));
}

void TabelaObjetosWidget::definirConfiguracaoBotoes(unsigned conf_botoes)
{
 bool btn_mover, btn_edt, btn_ins, btn_limpar, btn_rem, btn_atual;

 /* Verificando, atrav√©s de opera√ß√µes bit a bit, a presen√ßa de cada tipo de
    bot√£o na configura√ß√£o passada */
 btn_mover=(conf_botoes & BTN_MOVER_ITENS) == BTN_MOVER_ITENS;
 btn_edt=(conf_botoes & BTN_EDITAR_ITEM) == BTN_EDITAR_ITEM;
 btn_ins=(conf_botoes & BTN_INSERIR_ITEM) == BTN_INSERIR_ITEM;
 btn_rem=(conf_botoes & BTN_REMOVER_ITEM) == BTN_REMOVER_ITEM;
 btn_limpar=(conf_botoes & BTN_LIMPAR_ITENS) == BTN_LIMPAR_ITENS;
 btn_atual=(conf_botoes & BTN_ATUALIZAR_ITEM) == BTN_ATUALIZAR_ITEM;

 //Seta os bt√µes vis√≠veis de acordo com a presen√ßa dos mesmos na configura√ß√£o de bot√µes
 mover_baixo_tb->setVisible(btn_mover);
 mover_cima_tb->setVisible(btn_mover);
 mover_primeiro_tb->setVisible(btn_mover);
 mover_ultimo_tb->setVisible(btn_mover);

 editar_tb->setVisible(btn_edt);
 remover_todas_tb->setVisible(btn_limpar);

 adicionar_tb->setVisible(btn_ins);
 remover_tb->setVisible(btn_rem);
 atualizar_tb->setVisible(btn_atual);

 if(!btn_edt && !btn_mover && !btn_ins && !btn_limpar && !btn_rem && !btn_atual)
 {
  tabelaobj_grid->removeWidget(tabela_tbw);
  tabelaobj_grid->addWidget(tabela_tbw,0,0,1,10);
 }
}

void TabelaObjetosWidget::definirNumColunas(unsigned num_colunas)
{
 if(num_colunas > 0)
 {
  unsigned i;
  QTableWidgetItem *item=NULL;

  /* A adi√ß√£o dos r√≥tulos de cabe√ßalhos das colunas se iniciar√°
     a partir da ultima coluna */
  i=tabela_tbw->columnCount();
  //Define o novo n√∫mero de colunas da tabela
  tabela_tbw->setColumnCount(num_colunas);

  /* Adiciona novos r√≥tulos caso o n√∫mero de colunas do par√¢metro
     seja superior ao n√∫mero de colunas atual da tabela */
  for(;i < num_colunas; i++)
  {
   item=new QTableWidgetItem;
   tabela_tbw->setHorizontalHeaderItem(static_cast<int>(i),item);
  }
 }
}

void TabelaObjetosWidget::definirRotuloCabecalho(const QString &rotulo, unsigned idx_col)
{
 QTableWidgetItem *item=NULL;

 /* Caso o √≠ndice da coluna do cabe√ßalho referenciada seja inv√°lida
    dispara um erro indicando a situa√ß√£o */
 if(idx_col >= static_cast<unsigned>(tabela_tbw->columnCount()))
  throw Excecao(ERR_PGMODELERUI_REFCOLTABOBJIDXINV,__PRETTY_FUNCTION__,__FILE__,__LINE__);

 //Obt√©m o item que representa o cabe√ßalho na coluna indicada
 item=tabela_tbw->horizontalHeaderItem(idx_col);
 //Define o r√≥tulo do item
 item->setText(rotulo);
}

void TabelaObjetosWidget::definirIconeCabecalho(const QIcon &icone, unsigned idx_col)
{
 QTableWidgetItem *item=NULL;

 /* Caso o √≠ndice da coluna do cabe√ßalho referenciada seja inv√°lida
    dispara um erro indicando a situa√ß√£o */
 if(idx_col >= static_cast<unsigned>(tabela_tbw->columnCount()))
  throw Excecao(ERR_PGMODELERUI_REFCOLTABOBJIDXINV,__PRETTY_FUNCTION__,__FILE__,__LINE__);

 //Obt√©m o item que representa o cabe√ßalho na coluna indicada
 item=tabela_tbw->horizontalHeaderItem(idx_col);
 //Define o √≠cone do item obtido
 item->setIcon(icone);
}

void TabelaObjetosWidget::definirIconeCelula(const QIcon &icone, unsigned idx_lin, unsigned idx_col)
{
 QTableWidgetItem *item=NULL;

 /* Caso o √≠ndice da linha da c√©lula a ser referenciada seja inv√°lida
    dispara um erro indicando a situa√ß√£o */
 if(idx_lin >= static_cast<unsigned>(tabela_tbw->rowCount()))
  throw Excecao(ERR_PGMODELERUI_REFLINTABOBJIDXINV,__PRETTY_FUNCTION__,__FILE__,__LINE__);

 /* Caso o √≠ndice da coluna da c√©lula a ser referenciada seja inv√°lida
    dispara um erro indicando a situa√ß√£o */
 if(idx_col >= static_cast<unsigned>(tabela_tbw->columnCount()))
  throw Excecao(ERR_PGMODELERUI_REFCOLTABOBJIDXINV,__PRETTY_FUNCTION__,__FILE__,__LINE__);

 //Obt√©m o item o qual representa a c√©lula
 item=tabela_tbw->item(idx_lin, idx_col);
 //Define o √≠cone da c√©lula
 item->setIcon(icone);
}

void TabelaObjetosWidget::definirTextoCelula(const QString &texto, unsigned idx_lin, unsigned idx_col)
{
 QTableWidgetItem *item=NULL;

 /* Caso o √≠ndice da linha da c√©lula a ser referenciada seja inv√°lida
    dispara um erro indicando a situa√ß√£o */
 if(idx_lin >= static_cast<unsigned>(tabela_tbw->rowCount()))
  throw Excecao(ERR_PGMODELERUI_REFLINTABOBJIDXINV,__PRETTY_FUNCTION__,__FILE__,__LINE__);

 /* Caso o √≠ndice da coluna da c√©lula a ser referenciada seja inv√°lida
    dispara um erro indicando a situa√ß√£o */
 if(idx_col >= static_cast<unsigned>(tabela_tbw->columnCount()))
  throw Excecao(ERR_PGMODELERUI_REFCOLTABOBJIDXINV,__PRETTY_FUNCTION__,__FILE__,__LINE__);

 //Obt√©m o item o qual representa a c√©lula
 item=tabela_tbw->item(idx_lin,idx_col);
 //Define o texto da c√©lula
 item->setText(texto);
}

void TabelaObjetosWidget::definirFonteLinha(int idx_lin, const QFont &fonte, const QColor &cor_texto, const QColor &cor_fundo)
{
 QTableWidgetItem *item=NULL;
 int qtd_col, i;

 /* Caso o √≠ndice da linha da c√©lula a ser referenciada seja inv√°lida
    dispara um erro indicando a situa√ß√£o */
 if(idx_lin >= tabela_tbw->rowCount())
  throw Excecao(ERR_PGMODELERUI_REFLINTABOBJIDXINV,__PRETTY_FUNCTION__,__FILE__,__LINE__);

 qtd_col=tabela_tbw->columnCount();
 for(i=0; i < qtd_col; i++)
 {
  item=tabela_tbw->item(idx_lin, i);
  item->setFont(fonte);
  item->setForeground(cor_texto);
  item->setBackgroundColor(cor_fundo);
 }
}

void TabelaObjetosWidget::definirDadoLinha(const QVariant &dado, unsigned idx_lin)
{
 QTableWidgetItem *item=NULL;

 /* Caso o √≠ndice da linha da c√©lula a ser referenciada seja inv√°lida
    dispara um erro indicando a situa√ß√£o */
 if(idx_lin >= static_cast<unsigned>(tabela_tbw->rowCount()))
  throw Excecao(ERR_PGMODELERUI_REFLINTABOBJIDXINV,__PRETTY_FUNCTION__,__FILE__,__LINE__);

 /* Obt√©m o √≠tem que representa o cabe√ßalho vertical da linha.
    Este cabe√ßalho √© usado para armazenar o dado que a linha guarda.
    Os dados s√£o guardados em forma de um QVariant. */
 item=tabela_tbw->verticalHeaderItem(idx_lin);
 //Armazena dentro do item o dado passado no par√¢metro
 item->setData(Qt::UserRole, dado);
}

unsigned TabelaObjetosWidget::obterNumColunas(void)
{
 //Retorna o n√∫mero de colunas do objeto tabela
 return(tabela_tbw->columnCount());
}

unsigned TabelaObjetosWidget::obterNumLinhas(void)
{
 //Retorna o n√∫mero de linhas do objeto tabela
 return(tabela_tbw->rowCount());
}

QString TabelaObjetosWidget::obterRotuloCabecalho(unsigned idx_col)
{
 QTableWidgetItem *item=NULL;

 /* Caso o √≠ndice da coluna do cabe√ßalho a ser referenciado seja inv√°lida
    dispara um erro indicando a situa√ß√£o */
 if(idx_col >= static_cast<unsigned>(tabela_tbw->columnCount()))
  throw Excecao(ERR_PGMODELERUI_REFCOLTABOBJIDXINV,__PRETTY_FUNCTION__,__FILE__,__LINE__);

 //Obt√©m o cabe√ßalho e retorna seu texto
 item=tabela_tbw->horizontalHeaderItem(idx_col);
 return(item->text());
}

QString TabelaObjetosWidget::obterTextoCelula(unsigned idx_lin, unsigned idx_col)
{
 QTableWidgetItem *item=NULL;

 /* Caso o √≠ndice da linha da c√©lula a ser referenciada seja inv√°lida
    dispara um erro indicando a situa√ß√£o */
 if(idx_lin >= static_cast<unsigned>(tabela_tbw->rowCount()))
  throw Excecao(ERR_PGMODELERUI_REFLINTABOBJIDXINV,__PRETTY_FUNCTION__,__FILE__,__LINE__);

 /* Caso o √≠ndice da coluna da c√©lula a ser referenciada seja inv√°lida
    dispara um erro indicando a situa√ß√£o */
 if(idx_col >= static_cast<unsigned>(tabela_tbw->columnCount()))
  throw Excecao(ERR_PGMODELERUI_REFCOLTABOBJIDXINV,__PRETTY_FUNCTION__,__FILE__,__LINE__);

 /* Obt√©m o item que representa a c√©lula na linha e coluna especificadas
    e procede com o retorno do seu texto */
 item=tabela_tbw->item(idx_lin,idx_col);
 return(item->text());
}

QVariant TabelaObjetosWidget::obterDadoLinha(unsigned idx_lin)
{
 QTableWidgetItem *item=NULL;

 /* Caso o √≠ndice da linha a ser referenciada seja inv√°lida
    dispara um erro indicando a situa√ß√£o */
 if(idx_lin >= static_cast<unsigned>(tabela_tbw->rowCount()))
  throw Excecao(ERR_PGMODELERUI_REFLINTABOBJIDXINV,__PRETTY_FUNCTION__,__FILE__,__LINE__);

 //Obt√©m o cabe√ßalho vertical da linha o qual armazena o dado desta
 item=tabela_tbw->verticalHeaderItem(idx_lin);
 //Retorna o dado armazenado pelo item
 return(item->data(Qt::UserRole));
}

int TabelaObjetosWidget::obterLinhaSelecionada(void)
{
 return(tabela_tbw->currentRow());
}

int TabelaObjetosWidget::obterIndiceLinha(const QVariant &dado)
{
 unsigned i, qtd;
 QTableWidgetItem *item=NULL;
 bool enc=false;
 QVariant dado_aux;

 //Obt√©m a quantidade de linhas da tabela
 qtd=tabela_tbw->rowCount();

 /* Varre as linhas da tabela comparando o dado armazenado
    em cada uma delas com o dado vindo do par√¢metro */
 for(i=0; !enc && i < qtd; i++)
 {
  //Obt√©m o item que armazena o dado da linha
  item=tabela_tbw->verticalHeaderItem(i);

  //Faz a compara√ß√£o entre os valores do mesmo e do par√¢metro
  enc=(item &&
       item->data(Qt::UserRole).value<void *>() == dado.value<void *>());
 }

 /* Caso o mesmo n√£o seja encontrado retorna -1
    caso contr√°rio retorna o √≠ndice da linha onde
    o mesmo se encontra */
 if(!enc)
  return(-1);
 else
  return(i);
}

void TabelaObjetosWidget::adicionarColuna(unsigned idx_col)
{
 /* Caso o √≠ndice da coluna anterio√  coluna a ser adcionada seja inv√°lido
    adiciona a coluna ao final da lista de colunas */
 if(idx_col >= static_cast<unsigned>(tabela_tbw->columnCount()))
  idx_col=tabela_tbw->columnCount();

 //Adiciona a coluna
 tabela_tbw->insertColumn(idx_col);
 tabela_tbw->clearSelection();
 habilitarBotoes();

 //Emite um sinal indicando em qual √≠ndice foi adicionada uma coluna
 emit s_colunaAdicionada(idx_col);
}

void TabelaObjetosWidget::selecionarLinha(int idx_lin)
{
 QTableWidgetItem *item=NULL;

 item=tabela_tbw->item(0,idx_lin);

 if(item)
 {
  item=tabela_tbw->item(idx_lin,0);
  item->setSelected(true);
  tabela_tbw->setCurrentItem(item);
  habilitarBotoes();
 }
}

void TabelaObjetosWidget::adicionarLinha(void)
{
 QTableWidgetItem *item=NULL;
 unsigned i, lin, qtd_col=tabela_tbw->columnCount();

 //A nova linha ser√° inserida ap√≥s a √∫ltima linha
 lin=tabela_tbw->rowCount();
 tabela_tbw->insertRow(lin);

 //Cria o cabe√ßalho vertical da linha adicionada
 item=new QTableWidgetItem;
 item->setText(QString("%1").arg(lin+1));
 tabela_tbw->setVerticalHeaderItem(lin,item);

 /* Ap√≥s inserida a linha as colunas da nova linha
    precisam ser criadas */
 for(i=0; i < qtd_col; i++)
 {
  //Cria um item que representa a coluna atual da nova linha
  item=new QTableWidgetItem;
  //Insere o item criado na coluna atual (i)
  tabela_tbw->setItem(lin,i,item);
 }

 //Seleciona os itens da linha inserida
 item=tabela_tbw->item(lin,0);
 item->setSelected(true);
 tabela_tbw->setCurrentItem(item);

 /* Executa o m√©todo de habilita√ß√£o dos bot√µes de acordo
    com a linha atual */
 habilitarBotoes();
 //Emite um sinal com o √≠ndice da linha adicionada
 emit s_linhaAdicionada(lin);
}

void TabelaObjetosWidget::removerLinha(unsigned idx_lin)
{
 unsigned i, qtd;
 bool conf;

 /* Caso o √≠ndice da linha da c√©lula a ser referenciada seja inv√°lida
    dispara um erro indicando a situa√ß√£o */
 if(idx_lin >= static_cast<unsigned>(tabela_tbw->rowCount()))
  throw Excecao(ERR_PGMODELERUI_REFLINTABOBJIDXINV,__PRETTY_FUNCTION__,__FILE__,__LINE__);

 //Limpa a sele√ß√£o da linha atual
 tabela_tbw->clearSelection();
 qtd=tabela_tbw->columnCount();

 //Seleciona todas a colunas da linha a ser removida
 for(i=0; i < qtd; i++)
  tabela_tbw->item(idx_lin, i)->setSelected(true);

 //Marca a linha cujo √≠ndice √© idx_lin como sendo a linha atual da tabela
 tabela_tbw->setCurrentItem(tabela_tbw->item(idx_lin,0));

 conf=conf_exclusoes;
 conf_exclusoes=false;
 //Remove a linha selecionada
 removerLinha();
 conf_exclusoes=conf;
}

void TabelaObjetosWidget::removerLinha(void)
{
 /* Caso haja alguma linha selecionada ou seja o √≠ndice
    da linha atual seja igual ou superior a zero */
 if(tabela_tbw->currentRow()>=0)
 {
  unsigned /*qtd_lin,*/
           idx_lin=tabela_tbw->currentRow();
  //Obt√©m o item selecionado na linha atual
  QTableWidgetItem *item=tabela_tbw->currentItem();

  if(item->isSelected())
  {
   if(conf_exclusoes)
    caixa_msg->show(trUtf8("Confirma√ß√£o"),trUtf8("Do you really want to remove the selected item?"),
                    CaixaMensagem::ICONE_CONFIRM, CaixaMensagem::BOTAO_SIM_NAO);

   if(!conf_exclusoes || (conf_exclusoes && caixa_msg->result()==QDialog::Accepted))
   {
    //Remove a linha atual
    tabela_tbw->removeRow(idx_lin);
    tabela_tbw->setCurrentItem(NULL);
    habilitarBotoes();

    //Emite o sinal de linha removida com o √≠ndice da linha exclu√≠da
    emit s_linhaRemovida(idx_lin);
   }
  }
 }
}

void TabelaObjetosWidget::removerLinhas(void)
{
 if(tabela_tbw->rowCount() > 0)
 {
  QObject *obj_sender=sender();

  /* A caixa de mensagem s√≥ √© exibida caso a confirma√ß√£o de exclus√£o esteja configurada
     e o objeto sender esteja setado para o bot√£o "remover todas".
     Este objeto sender indica que o m√©todo foi
     chamado atrav√©s do acionamento deste objeto. Este m√©todo pode ser chamado diretamente
     sem ser atrav√©s de outro objeto, desta forma, todos os itens s√£o removidos sem confirma√ß√£o
     independente da confirma√ß√£o estar configurada ou n√£o. Isto √© util no caso de uma limpeza
     da tabela para reuso. */
  if(conf_exclusoes && obj_sender==remover_todas_tb)
   caixa_msg->show(trUtf8("Confirma√ß√£o"),trUtf8("Do you really want to remove the all items?"),
                   CaixaMensagem::ICONE_CONFIRM, CaixaMensagem::BOTAO_SIM_NAO);

  if(!conf_exclusoes || (conf_exclusoes && obj_sender!=remover_todas_tb) ||
    (conf_exclusoes &&  obj_sender==remover_todas_tb && caixa_msg->result()==QDialog::Accepted))
  {
   //Remove as linhas enquanto a quantidade n√£o atinge zero
   while(tabela_tbw->rowCount() > 0)
    tabela_tbw->removeRow(0);

   habilitarBotoes();

   // Emite o sinal indicando que as linhas da tabela foram removidas caso
   emit s_linhasRemovidas();
  }
 }
}

void TabelaObjetosWidget::removerColuna(unsigned idx_col)
{
 /* Caso o √≠ndice da coluna a ser removida seja inv√°lido retorna um erro
    e aborta a opera√ß√£o */
 if(idx_col >= static_cast<unsigned>(tabela_tbw->columnCount()))
  throw Excecao(ERR_PGMODELERUI_REFCOLTABOBJIDXINV,__PRETTY_FUNCTION__,__FILE__,__LINE__);

 //Remove a coluna no indice especificado e limpa a sele√ß√£o
 tabela_tbw->removeColumn(idx_col);
 tabela_tbw->clearSelection();
 habilitarBotoes();
 //Emite o sinal indicando a coluna removida
 emit s_colunaRemovida(idx_col);
}

void TabelaObjetosWidget::moverLinhas(void)
{
 QObject *obj_sender=sender();
 QTableWidgetItem *item=NULL, *item1=NULL;
 int lin=-1, lin1=-1;
 unsigned col, qtd_col=tabela_tbw->columnCount();
 QVariant dado_aux;

 /* Obt√©m o √≠ndice da linha atual. √â com base nesse √≠ndice
    que ser√° verificado com qual linha ser√° trocada a posi√ß√£o
    da linha selecionada */
 lin=tabela_tbw->currentRow();

 /* Caso o bot√£o de mover uma linha para baixo for acionado
    a linha selecionada ter√° sua posi√ß√£o trocada com a linha
    imediatamente abaixo dela */
 if(obj_sender==mover_baixo_tb)
  lin1=lin+1;
  /* Caso o bot√£o de mover uma linha para cima for acionado
    a linha selecionada ter√° sua posi√ß√£o trocada com a linha
    imediatamente acima dela */
 else if(obj_sender==mover_cima_tb)
  lin1=lin-1;
 /* Caso o bot√£o de mover para a primeira linha for acionado
    a linha selecionada ter√° sua posi√ß√£o trocada com a
    primeira linha da tabela */
 else if(obj_sender==mover_primeiro_tb)
  lin1=0;
 /* Caso o bot√£o de mover para a √∫ltima linha for acionado
    a linha selecionada ter√° sua posi√ß√£o trocada com a
    √∫ltima linha da tabela */
 else if(obj_sender==mover_ultimo_tb)
  lin1=tabela_tbw->rowCount()-1;
 /* Verifica se os √≠ndices da linhas a serem trocadas s√£o v√°lidos
    ou seja, s√£o diferentes entre si e n√£o ultrapassam o limite de
    linhas presentes na tabela */
 if(lin >= 0 && lin < tabela_tbw->rowCount() &&
    lin1 >= 0 && lin1 < tabela_tbw->rowCount() &&
    lin != lin1)
 {
  /* Para se trocar a posi√ß√£o da linhas √© necess√°rio trocar coluna por coluna
     desta forma o for a seguir executa a troca de cada coluna das linhas
     envolvida */
  for(col=0; col < qtd_col; col++)
  {
   /* Obt√©m uma coluna da linha atual e em seguida remove
      a mesma sem desaloc√°-la */
   item=tabela_tbw->item(lin, col);
   tabela_tbw->takeItem(lin, col);

  /* Obt√©m uma coluna da linha que ser√° trocada com atual
     e em seguida remove a mesma sem desaloc√°-la */
   item1=tabela_tbw->item(lin1, col);
   tabela_tbw->takeItem(lin1, col);

   /* Efetiva a troca das linhas, onde o item obtido da linha atual
      passa a pertencer a linha vizinha e o item da linha vizinha
      passa a fazer parte da linha atual */
   tabela_tbw->setItem(lin, col, item1);
   tabela_tbw->setItem(lin1, col, item);

   item1->setSelected(false);
   item->setSelected(true);
  }

  tabela_tbw->setCurrentItem(item);

  /* Fazendo a troca dos dados das linhas. Para isso
   √© necess√°rio obter os cabe√ßalhos verticais os quais
   guardam a informa√ß√£o em si da linha da tabela */
  item=tabela_tbw->verticalHeaderItem(lin);
  item1=tabela_tbw->verticalHeaderItem(lin1);

  if(item && item1)
  {
   //Obtido os items faz a troca dos dados entre eles
   dado_aux=item->data(Qt::UserRole);
   item->setData(Qt::UserRole, item1->data(Qt::UserRole));
   item1->setData(Qt::UserRole, dado_aux);
  }

  habilitarBotoes();
  emit s_linhasMovidas(lin, lin1);
 }
}

void TabelaObjetosWidget::editarLinha(void)
{
 /* Para este m√©todo nada √© executado apenas um sinal √© emitido
    com o √≠ndice da linha a ser editada. Quem deve tratar a edi√ß√£o
    da linha √© o objeto externo o qual faz uso da tabela. */
 emit s_linhaEditada(tabela_tbw->currentRow());
}

void TabelaObjetosWidget::atualizarLinha(void)
{
 /* Para este m√©todo nada √© executado apenas um sinal √© emitido
    com o √≠ndice da linha a ser editada. Quem deve tratar a edi√ß√£o
    da linha √© o objeto externo o qual faz uso da tabela. */
 emit s_linhaAtualizada(tabela_tbw->currentRow());
}

void TabelaObjetosWidget::limparSelecao(void)
{
 tabela_tbw->clearSelection();
 tabela_tbw->setCurrentItem(NULL);
 habilitarBotoes();
}

void TabelaObjetosWidget::habilitarBotoes(unsigned conf_botoes, bool valor)
{
 int lin=-1;
 //Obt√©m o item atual caso haja algum selecionado
 QTableWidgetItem *item=tabela_tbw->currentItem();

 //Obt√©m a linha a qual o item pertence
 if(item)
  lin=item->row();

 /* Atribui o valor booleano passado caso o bot√£o esteja presente
    na configura√ß√£o de bot√µes informada */
 if((conf_botoes & BTN_MOVER_ITENS) == BTN_MOVER_ITENS)
 {
  /* O bot√£o de mover uma linha para cima deve ser habilitado
     caso a linha selecionada n√£o seja a primeira linha */
  mover_cima_tb->setEnabled(valor && lin > 0);

  /* O bot√£o de mover uma linha para baixo deve ser habilitado
     caso a linha selecionada n√£o seja a √∫ltima linha */
  mover_baixo_tb->setEnabled(valor && lin >= 0 && lin < tabela_tbw->rowCount()-1);

   /* O bot√£o de mover para √∫ltima linha deve ser habilitado
     caso a linha selecionada n√£o seja a √∫ltima */
  mover_primeiro_tb->setEnabled(valor && lin > 0 && lin<=tabela_tbw->rowCount()-1);

  /* O bot√£o de mover para primeira linha deve ser habilitado
     caso a linha selecionada n√£o seja a primeira */
  mover_ultimo_tb->setEnabled(valor && lin >=0 && lin < tabela_tbw->rowCount()-1);
 }

 if((conf_botoes & BTN_EDITAR_ITEM) == BTN_EDITAR_ITEM)
  editar_tb->setEnabled(valor && lin >= 0);

 if((conf_botoes & BTN_INSERIR_ITEM) == BTN_INSERIR_ITEM)
  adicionar_tb->setEnabled(valor);

 if((conf_botoes & BTN_REMOVER_ITEM) == BTN_REMOVER_ITEM)
  remover_tb->setEnabled(valor && lin >= 0);

 if((conf_botoes & BTN_LIMPAR_ITENS) == BTN_LIMPAR_ITENS)
  remover_todas_tb->setEnabled(valor && tabela_tbw->rowCount() > 0);

 if((conf_botoes & BTN_ATUALIZAR_ITEM) == BTN_ATUALIZAR_ITEM)
  atualizar_tb->setEnabled(valor && lin >= 0);
}

void TabelaObjetosWidget::habilitarBotoes(void)
{
 //Obt√©m o item atual caso haja algum selecionado
 QTableWidgetItem *item=tabela_tbw->currentItem();

 habilitarBotoes(TODOS_BOTOES, true);

 /* Caso uma linha esteja selecionada emite o sinal indicativo de sele√ß√£o de linha,
    este sinal √© interessante quando se quer ter acesso diret√  linha selecionada
    sem ter que chamar o m√©todo de obterLinhaSelecionada() */
 if(item && item->row() >= 0)
  emit s_linhaSelecionada(item->row());
}

